From 8aa6935b068baf1da003ab5a0ffb856016a9ac5c Mon Sep 17 00:00:00 2001
From: wudihui <wudihui@xiaomi.corp-partner.google.com>
Date: Fri, 8 May 2026 20:23:06 +0800
Subject: [PATCH] ANDROID: GKI: add vendor hooks for Track direct_reclaim
 allocation latency.

Quantify the benefit through the time consumption change of this function.

Bug: 514559198
Bug: 506924601

Change-Id: I08f89ddfdd3f0e577ead0f0794f39ba4572f145d
Signed-off-by: wudihui <wudihui@xiaomi.corp-partner.google.com>
(cherry picked from commit fdfb13d4c0348e6fe45defb61de30b8be4f43923)
---
 drivers/android/vendor_hooks.c | 2 ++
 include/trace/hooks/mm.h       | 6 ++++++
 mm/page_alloc.c                | 3 +++
 3 files changed, 11 insertions(+)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index 6cc6df646e1b..d9149a4fe650 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -602,6 +602,8 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_compact_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_compact_finished);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_direct_reclaim_enter);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_direct_reclaim_exit);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_direct_reclaim_start);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_direct_reclaim_end);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_may_oom_exit);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_calculate_totalreserve_pages);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_madvise_cold_pageout_skip);
diff --git a/include/trace/hooks/mm.h b/include/trace/hooks/mm.h
index 37b59ad3d99c..3bfa5621534a 100644
--- a/include/trace/hooks/mm.h
+++ b/include/trace/hooks/mm.h
@@ -533,6 +533,12 @@ DECLARE_HOOK(android_vh_mm_direct_reclaim_enter,
 DECLARE_HOOK(android_vh_mm_direct_reclaim_exit,
 	TP_PROTO(unsigned long did_some_progress, int retry_times),
 	TP_ARGS(did_some_progress, retry_times));
+DECLARE_HOOK(android_vh_mm_direct_reclaim_start,
+       TP_PROTO(u64 *stime),
+       TP_ARGS(stime));
+DECLARE_HOOK(android_vh_mm_direct_reclaim_end,
+       TP_PROTO(unsigned int order, u64 stime),
+       TP_ARGS(order, stime));
 struct oom_control;
 DECLARE_HOOK(android_vh_mm_may_oom_exit,
 	TP_PROTO(struct oom_control *oc, unsigned long did_some_progress),
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 691898e785be..3da165f4e64a 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -4433,8 +4433,10 @@ __alloc_pages_direct_reclaim(gfp_t gfp_mask, unsigned int order,
 	unsigned long pflags;
 	bool drained = false;
 	bool skip_pcp_drain = false;
+	u64 stime = 0;
 
 	trace_android_vh_mm_direct_reclaim_enter(order);
+	trace_android_vh_mm_direct_reclaim_start(&stime);
 	psi_memstall_enter(&pflags);
 	*did_some_progress = __perform_reclaim(gfp_mask, order, ac);
 	if (unlikely(!(*did_some_progress)))
@@ -4461,6 +4463,7 @@ __alloc_pages_direct_reclaim(gfp_t gfp_mask, unsigned int order,
 out:
 	psi_memstall_leave(&pflags);
 	trace_android_vh_mm_direct_reclaim_exit(*did_some_progress, retry_times);
+	trace_android_vh_mm_direct_reclaim_end(order, stime);
 	return page;
 }
 
-- 
2.55.0

